TNutils.py

```diff
@@ -1000,4 +1002,87 @@ def reconstruct_SLOW(mps, corr_img):
                 reconstructed[site] = 1
                 
     return reconstructed
+
+def reconstruct(mps, corr_img):
+    
+    # Copy the tensor, we need to perform vertical
+    # contractions among all know pixels
+    rec_mps = copy.copy(mps)
+    rec_mps.normalize()
+    
+    # Contracting know pixels
+    for site, pixel in enumerate(corr_img):
+        if pixel == 0:
+            if site == 0 or site == len(rec_mps.tensors) - 1:
+                data = np.einsum('ab,b',rec_mps[site].data, [0,1] )
+                inds = tuple(list(mps.tensors[site].inds)[:-1])
+                rec_mps.tensors[site].modify(data=data, inds = inds)
+            else:
+                data = np.einsum('abc,c',rec_mps[site].data, [0,1] )
+                inds = tuple(list(mps.tensors[site].inds)[:-1])
+                rec_mps.tensors[site].modify(data=data, inds = inds)
+        elif pixel == 1:
+            if site == 0 or site == len(rec_mps.tensors) - 1:
+                data = np.einsum('ab,b',rec_mps[site].data, [1,0] )
+                inds = tuple(list(mps.tensors[site].inds)[:-1])
+                rec_mps.tensors[site].modify(data=data, inds = inds)
+            else:
+                data = np.einsum('abc,c',rec_mps[site].data, [1,0] )
+                inds = tuple(list(mps.tensors[site].inds)[:-1])
+                rec_mps.tensors[site].modify(data=data, inds = inds)
+                
+    first = False # check if we already found an unknown pixel
+    upixel = -1
+    for site in range(len(rec_mps.tensors)):
+        if site == 0:
+            ut = rec_mps.tensors[site]
+            if 'v0' in rec_mps.tensors[site].inds: # 0 is unknown
+                ut = tneinsum2(ut,rec_mps.tensors[site+1])
+            else:
+                ut = tneinsum2(ut, rec_mps.tensors[site+1])
+        else:
+            if 'v'+str(site) in rec_mps.tensors[site].inds:
+                upixel = upixel + 1
+                ut.modify(tags='U'+str(upixel))
+                if not first:
+                    utn = qtn.Tensor(ut)
+                else:
+                    utn = utn & ut
```

Comment:
Then it is being rebuilt as a product state; could it be better to keep it a list of tensors?